packages/react/src/combobox/positioner/ComboboxPositioner.tsx

```diff
@@ -0,0 +1,195 @@
+'use client';
+import * as React from 'react';
+import type { VirtualElement } from '@floating-ui/react-dom';
+import { useStore } from '@base-ui-components/utils/store';
+import { useEventCallback } from '@base-ui-components/utils/useEventCallback';
+import { inertValue } from '@base-ui-components/utils/inertValue';
+import {
+  useComboboxFloatingContext,
+  useComboboxRootContext,
+  useComboboxDerivedItemsContext,
+} from '../root/ComboboxRootContext';
+import { ComboboxPositionerContext } from './ComboboxPositionerContext';
+import { type Side, type Align, useAnchorPositioning } from '../../utils/useAnchorPositioning';
+import type { BaseUIComponentProps, HTMLProps } from '../../utils/types';
+import { popupStateMapping } from '../../utils/popupStateMapping';
+import { useComboboxPortalContext } from '../portal/ComboboxPortalContext';
+import { DROPDOWN_COLLISION_AVOIDANCE } from '../../utils/constants';
+import { useRenderElement } from '../../utils/useRenderElement';
+import { selectors } from '../store';
+import { useComboboxDefaultAnchor } from './ComboboxDefaultAnchorContext';
+import { InternalBackdrop } from '../../utils/InternalBackdrop';
+import { useScrollLock } from '../../utils/useScrollLock';
+
+/**
+ * Positions the popup against the trigger.
+ * Renders a `<div>` element.
+ *
+ * Documentation: [Base UI Preview Card](https://base-ui.com/react/components/preview-card)
+ */
+export const ComboboxPositioner = React.forwardRef(function ComboboxPositioner(
+  componentProps: ComboboxPositioner.Props,
+  forwardedRef: React.ForwardedRef<HTMLDivElement>,
+) {
+  const {
+    render,
+    className,
+    anchor,
+    positionMethod = 'absolute',
+    side = 'bottom',
+    align = 'center',
+    sideOffset = 0,
+    alignOffset = 0,
+    collisionBoundary = 'clipping-ancestors',
+    collisionPadding = 5,
+    arrowPadding = 5,
+    sticky = false,
+    trackAnchor = true,
+    collisionAvoidance = DROPDOWN_COLLISION_AVOIDANCE,
+    ...elementProps
+  } = componentProps;
+
+  const { store, modal } = useComboboxRootContext();
+  const { filteredItems } = useComboboxDerivedItemsContext();
+  const floatingRootContext = useComboboxFloatingContext();
+  const keepMounted = useComboboxPortalContext();
+
+  const open = useStore(store, selectors.open);
+  const mounted = useStore(store, selectors.mounted);
+  const empty = filteredItems.length === 0;
+  const openMethod = useStore(store, selectors.openMethod);
+
+  const defaultAnchor = useComboboxDefaultAnchor();
+  const triggerElement = useStore(store, selectors.triggerElement);
+  const inputElement = useStore(store, selectors.inputElement);
+
+  const resolvedAnchor = React.useMemo(() => {
+    if (anchor != null) {
+      return anchor;
+    }
+    if (defaultAnchor === 'trigger') {
+      return triggerElement;
+    }
+    return inputElement;
+  }, [anchor, defaultAnchor, triggerElement, inputElement]);
```

Comment:
The computation that is wrapped in `useMemo` is more expensive than the overhead of creating the allocations (closure, deps array) and checking the dependencies. I think you're overusing `useMemo`, it should be reserved for either non-primitive values that need to be stable, or computationally expensive operations. Calling `useMemo` is not a free operation - on a slow phone, it's 2-4ms a pop.

Reply:
I was tending to use it as an IIFE for the `const` sometimes, rather than calculating it externally, but since it's not free I'll do the calculations in the component body itself